File: apps/android/app/src/main/java/com/xueqing/app/infrastructure/deployment/DeploymentProfile.kt
package com.xueqing.app.infrastructure.deployment

import java.net.URI
import java.util.Locale
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

data class DeploymentProfileSource(
    val repository: String,
    val commit: String,
)

data class DeploymentProfile(
    val profileId: String,
    val environmentId: String,
    val trustDomainId: String,
    val providerId: String,
    val projectOrigin: String,
    val publishableKey: String,
    val requiredEdgeRegion: String,
    val capabilities: List<String>,
    val source: DeploymentProfileSource,
)

object DeploymentProfileParser {
    private const val CONTRACT = "deployment_profile_v1"
    private const val REPOSITORY = "qbjsdsb/xueqing-native"

    private val topKeys = setOf(
        "contract",
        "profile_id",
        "environment_id",
        "trust_domain_id",
        "provider_id",
        "project_origin",
        "publishable_key",
        "required_edge_region",
        "capabilities",
        "source",
    )
    private val sourceKeys = setOf("repository", "commit")
    private val supportedCapabilities = setOf(
        "auth",
        "database-rpc",
        "edge-functions",
        "private-storage",
    )
    private val identifierRegex = Regex("^[a-z0-9][a-z0-9._-]{1,95}$")
    private val regionRegex = Regex("^[a-z]{2}-[a-z0-9-]+-[0-9]+$")
    private val commitRegex = Regex("^[0-9a-f]{40}$")
    private val ipv4Regex = Regex("^(?:[0-9]{1,3}\\.){3}[0-9]{1,3}$")

    fun parse(value: String): DeploymentProfile {
        val root = try {
            Json.parseToJsonElement(value) as? JsonObject
                ?: error("Deployment profile must be a JSON object.")
        } catch (error: Exception) {
            throw IllegalArgumentException("Deployment profile JSON is malformed.", error)
        }

        require(root.keys == topKeys) {
            "Deployment profile keys do not match the accepted v1 contract."
        }
        require(root.requiredString("contract") == CONTRACT) {
            "Unsupported deployment profile contract."
        }

        val profileId = root.identifier("profile_id", 64)
        val environmentId = root.identifier("environment_id", 64)
        val trustDomainId = root.identifier("trust_domain_id", 96)
        val providerId = root.identifier("provider_id", 32)
        val origin = validateOrigin(root.requiredString("project_origin"))
        val publishableKey = validatePublishableKey(
            root.requiredString("publishable_key"),
        )
        val edgeRegion = root.requiredString("required_edge_region")
        require(regionRegex.matches(edgeRegion)) {
            "Deployment profile Edge region is invalid."
        }

        val capabilitiesElement = root["capabilities"] as? JsonArray
            ?: error("Deployment profile capabilities must be an array.")
        val capabilities = capabilitiesElement.map { element ->
            val primitive = element as? JsonPrimitive
                ?: error("Deployment profile capability must be a string.")
            require(primitive.isString)
            primitive.content.also { require(it.isNotBlank()) }
        }
        require(capabilities.isNotEmpty())
        require(capabilities.distinct().size == capabilities.size)
        require(capabilities == capabilities.sorted())
        require(capabilities.all(supportedCapabilities::contains))

        val source = root["source"] as? JsonObject
            ?: error("Deployment profile source must be an object.")
        require(source.keys == sourceKeys)
        val repository = source.requiredString("repository")
        require(repository == REPOSITORY)
        val commit = source.requiredString("commit")
        require(commitRegex.matches(commit))

        return DeploymentProfile(
            profileId = profileId,
            environmentId = environmentId,
            trustDomainId = trustDomainId,
            providerId = providerId,
            projectOrigin = origin,
            publishableKey = publishableKey,
            requiredEdgeRegion = edgeRegion,
            capabilities = capabilities,
            source = DeploymentProfileSource(repository, commit),
        )
    }

    private fun JsonObject.requiredString(name: String): String {
        val primitive = this[name] as? JsonPrimitive
            ?: error("Missing deployment profile string: $name")
        require(primitive.isString)
        return primitive.content.also { require(it.isNotBlank()) }
    }

    private fun JsonObject.identifier(name: String, maxLength: Int): String =
        requiredString(name).also { value ->
            require(value.length in 2..maxLength && identifierRegex.matches(value)) {
                "Deployment profile identifier '$name' is invalid."
            }
        }

    private fun validateOrigin(value: String): String {
        val uri = runCatching { URI.create(value) }.getOrNull()
            ?: error("Deployment profile project origin is invalid.")
        val host = uri.host?.lowercase(Locale.ROOT)
        require(
            uri.scheme == "https" &&
                !host.isNullOrBlank() &&
                uri.rawUserInfo == null &&
                uri.rawPath.isNullOrEmpty() &&
                uri.rawQuery == null &&
                uri.rawFragment == null &&
                uri.port in listOf(-1, 443) &&
                host !in setOf("localhost", "localhost.localdomain") &&
                !host.endsWith(".local") &&
                !host.endsWith(".internal") &&
                !host.endsWith(".localhost") &&
                !ipv4Regex.matches(host) &&
                !host.contains(":"),
        ) {
            "Deployment profile project origin is invalid."
        }
        return "${uri.scheme}://${uri.host}"
    }

    private fun validatePublishableKey(value: String): String {
        require(value.length in 16..2048 && value.none(Char::isWhitespace))
        val privilegedPrefix = listOf("sb_", "secret", "_").joinToString("")
        val privilegedRoleMarker = listOf("service", "_role").joinToString("")
        require(!value.contains(privilegedPrefix, ignoreCase = true))
        require(!value.contains(privilegedRoleMarker, ignoreCase = true))
        require(!value.contains("postgresql://", ignoreCase = true))
        require(!value.contains("postgres://", ignoreCase = true))
        return value
    }
}

File: apps/windows/src/Xueqing.Windows.Infrastructure/Deployment/DeploymentProfile.cs
using System.Net;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Xueqing.Windows.Infrastructure.Deployment;

public sealed record DeploymentProfileSource(
    string Repository,
    string Commit);

public sealed record DeploymentProfile(
    string ProfileId,
    string EnvironmentId,
    string TrustDomainId,
    string ProviderId,
    Uri ProjectOrigin,
    string PublishableKey,
    string RequiredEdgeRegion,
    IReadOnlyList<string> Capabilities,
    DeploymentProfileSource Source);

public static partial class DeploymentProfileParser
{
    private static readonly HashSet<string> SupportedCapabilities =
    [
        "auth",
        "database-rpc",
        "edge-functions",
        "private-storage",
    ];

    private static readonly string[] TopLevelKeys =
    [
        "contract",
        "profile_id",
        "environment_id",
        "trust_domain_id",
        "provider_id",
        "project_origin",
        "publishable_key",
        "required_edge_region",
        "capabilities",
        "source",
    ];

    private static readonly string[] SourceKeys =
    [
        "repository",
        "commit",
    ];

    public static DeploymentProfile Parse(string json)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(json);

        try
        {
            using var document = JsonDocument.Parse(json);
            var root = document.RootElement;
            RequireObject(root, "deployment profile");
            RequireExactKeys(root, TopLevelKeys, "deployment profile");

            if (RequiredString(root, "contract") != "deployment_profile_v1")
            {
                throw Invalid("Unsupported deployment profile contract.");
            }

            var profileId = Identifier(root, "profile_id", 64);
            var environmentId = Identifier(root, "environment_id", 64);
            var trustDomainId = Identifier(root, "trust_domain_id", 96);
            var providerId = Identifier(root, "provider_id", 32);
            var projectOrigin = ParseOrigin(RequiredString(root, "project_origin"));
            var publishableKey = ValidatePublishableKey(
                RequiredString(root, "publishable_key"));
            var requiredEdgeRegion = RequiredString(root, "required_edge_region");
            if (!RegionRegex().IsMatch(requiredEdgeRegion))
            {
                throw Invalid("Deployment profile Edge region is invalid.");
            }

            var capabilities = ParseCapabilities(root);
            var sourceElement = root.GetProperty("source");
            RequireObject(sourceElement, "deployment profile source");
            RequireExactKeys(sourceElement, SourceKeys, "deployment profile source");
            var repository = RequiredString(sourceElement, "repository");
            if (repository != "qbjsdsb/xueqing-native")
            {
                throw Invalid("Deployment profile source repository is invalid.");
            }

            var commit = RequiredString(sourceElement, "commit");
            if (!CommitRegex().IsMatch(commit))
            {
                throw Invalid("Deployment profile source commit is invalid.");
            }

            return new DeploymentProfile(
                profileId,
                environmentId,
                trustDomainId,
                providerId,
                projectOrigin,
                publishableKey,
                requiredEdgeRegion,
                capabilities,
                new DeploymentProfileSource(repository, commit));
        }
        catch (JsonException error)
        {
            throw new InvalidDataException(
                "Deployment profile JSON is malformed.",
                error);
        }
    }

    private static IReadOnlyList<string> ParseCapabilities(JsonElement root)
    {
        if (!root.TryGetProperty("capabilities", out var element) ||
            element.ValueKind != JsonValueKind.Array)
        {
            throw Invalid("Deployment profile capabilities must be an array.");
        }

        var values = element.EnumerateArray()
            .Select(value =>
                value.ValueKind == JsonValueKind.String &&
                !string.IsNullOrWhiteSpace(value.GetString())
                    ? value.GetString()!
                    : throw Invalid("Deployment profile capability is invalid."))
            .ToArray();

        if (values.Length == 0 ||
            values.Distinct(StringComparer.Ordinal).Count() != values.Length ||
            !values.SequenceEqual(values.Order(StringComparer.Ordinal), StringComparer.Ordinal) ||
            values.Any(value => !SupportedCapabilities.Contains(value)))
        {
            throw Invalid("Deployment profile capabilities are invalid.");
        }

        return values;
    }

    private static Uri ParseOrigin(string value)
    {
        if (!Uri.TryCreate(value, UriKind.Absolute, out var uri) ||
            uri.Scheme != Uri.UriSchemeHttps ||
            !string.IsNullOrEmpty(uri.UserInfo) ||
            uri.AbsolutePath != "/" ||
            !string.IsNullOrEmpty(uri.Query) ||
            !string.IsNullOrEmpty(uri.Fragment) ||
            uri.Port is not (-1 or 443) ||
            uri.IsLoopback ||
            uri.Host.EndsWith(".local", StringComparison.OrdinalIgnoreCase) ||
            uri.Host.EndsWith(".internal", StringComparison.OrdinalIgnoreCase) ||
            uri.Host.EndsWith(".localhost", StringComparison.OrdinalIgnoreCase) ||
            IPAddress.TryParse(uri.Host, out _))
        {
            throw Invalid("Deployment profile project origin is invalid.");
        }

        return new Uri(
            uri.GetLeftPart(UriPartial.Authority).TrimEnd('/'),
            UriKind.Absolute);
    }

    private static string ValidatePublishableKey(string value)
    {
        if (value.Length is < 16 or > 2048 ||
            value.Any(char.IsWhiteSpace))
        {
            throw Invalid("Deployment profile publishable key is invalid.");
        }

        var privilegedPrefix = string.Concat("sb_", "secret", "_");
        var privilegedRoleMarker = string.Concat("service", "_role");
        if (value.Contains(privilegedPrefix, StringComparison.OrdinalIgnoreCase) ||
            value.Contains(privilegedRoleMarker, StringComparison.OrdinalIgnoreCase) ||
            value.Contains("postgresql://", StringComparison.OrdinalIgnoreCase) ||
            value.Contains("postgres://", StringComparison.OrdinalIgnoreCase))
        {
            throw Invalid("Deployment profile contains privileged credential material.");
        }

        return value;
    }

    private static string Identifier(
        JsonElement root,
        string propertyName,
        int maxLength)
    {
        var value = RequiredString(root, propertyName);
        if (value.Length is < 2 ||
            value.Length > maxLength ||
            !IdentifierRegex().IsMatch(value))
        {
            throw Invalid($"Deployment profile identifier '{propertyName}' is invalid.");
        }

        return value;
    }

    private static string RequiredString(JsonElement root, string propertyName)
    {
        if (!root.TryGetProperty(propertyName, out var value) ||
            value.ValueKind != JsonValueKind.String ||
            string.IsNullOrWhiteSpace(value.GetString()))
        {
            throw Invalid($"Deployment profile property '{propertyName}' is missing/invalid.");
        }

        return value.GetString()!;
    }

    private static void RequireObject(JsonElement element, string label)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            throw Invalid($"{label} must be an object.");
        }
    }

    private static void RequireExactKeys(
        JsonElement element,
        IEnumerable<string> expected,
        string label)
    {
        var actual = element.EnumerateObject()
            .Select(property => property.Name)
            .ToHashSet(StringComparer.Ordinal);
        if (!actual.SetEquals(expected))
        {
            throw Invalid($"{label} keys do not match the accepted v1 contract.");
        }
    }

    private static InvalidDataException Invalid(string message) => new(message);

    [GeneratedRegex("^[a-z0-9][a-z0-9._-]{1,95}$", RegexOptions.CultureInvariant)]
    private static partial Regex IdentifierRegex();

    [GeneratedRegex("^[a-z]{2}-[a-z0-9-]+-[0-9]+$", RegexOptions.CultureInvariant)]
    private static partial Regex RegionRegex();

    [GeneratedRegex("^[0-9a-f]{40}$", RegexOptions.CultureInvariant)]
    private static partial Regex CommitRegex();
}
